File: app/src/main/java/com/bmn/bookfinder/activities/BookActivity.java
package com.bmn.bookfinder.activities;

import android.annotation.SuppressLint;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import com.bmn.bookfinder.R;
import com.bmn.bookfinder.data.room.AppDatabase;
import com.bmn.bookfinder.data.room.BookEntity;
import com.bmn.bookfinder.databinding.ActivityBookBinding;
import com.bumptech.glide.Glide;

import java.util.List;

public class BookActivity extends AppCompatActivity {

    private ActivityBookBinding binding;
    private BookEntity bookEntity;
    private String bookId;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityBookBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        bookId = BookActivityArgs.fromBundle(getIntent().getExtras()).getBookId();
        setBookData(bookId);
        binding.bookRating.setOnTouchListener((v, event) -> true);
        listenBookFavoriteChanges();
    }

    void listenBookFavoriteChanges() {
        binding.favoriteBook.setOnClickListener(v -> {
            bookEntity.setFavorite(!bookEntity.isFavorite());
            binding.favoriteBook.setImageDrawable(
                    bookEntity.isFavorite() ?
                            getDrawable(R.drawable.ic_heart_on) :
                            getDrawable(R.drawable.ic_heart)
            );
            AppDatabase.getDatabase(getApplicationContext()).getBookDao().setBookAsFavoriteById(bookId, bookEntity.isFavorite());

        });

    }

    private void setBookData(String bookId) {
        bookEntity = AppDatabase.getDatabase(getApplicationContext()).getBookDao().getBookById(bookId);
        formatAuthors(bookEntity.getAuthors());
        binding.bookDescription.setText(bookEntity.getDescription());
        binding.bookPageCount.setText(getString(R.string.page_count_d, bookEntity.getPageCount()));
        binding.bookPublishedDate.setText(bookEntity.getPublishedDate());
        binding.bookTitle.setText(bookEntity.getTitle());
        binding.bookRating.setRating((float) bookEntity.getAverageRating());
        binding.favoriteBook.setImageDrawable(
                bookEntity.isFavorite() ?
                        getDrawable(R.drawable.ic_heart_on) :
                        getDrawable(R.drawable.ic_heart)
        );

        Glide.with(this)
                .load(bookEntity.getThumbnailUrl())
                .into(binding.bookImage);

    }

    private void formatAuthors(List<String> authors) {
        StringBuilder authorsBuilder = new StringBuilder();

        for (String author : authors) {
            if (authors.indexOf(author) == authors.size() - 1) {
                authorsBuilder.append(author);
            }
            authorsBuilder.append(String.format("%s, ", author));
        }
        binding.bookAuthors.setText(authorsBuilder.toString());
    }
}

File: app/src/main/java/com/bmn/bookfinder/activities/BookActivity.kt
package com.bmn.bookfinder.activities

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.bmn.bookfinder.R
import com.bmn.bookfinder.data.room.AppDatabase
import com.bmn.bookfinder.data.room.BookEntity
import com.bmn.bookfinder.databinding.ActivityBookBinding
import com.bumptech.glide.Glide

class BookActivity : AppCompatActivity() {
    private var binding: ActivityBookBinding? = null
    private var bookEntity: BookEntity? = null
    private var bookId: String? = null

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBookBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        bookId = BookActivityArgs.fromBundle(intent.extras!!).bookId
        setBookData(bookId!!)
        binding!!.bookRating.setOnTouchListener { v: View?, event: MotionEvent? -> true }
        listenBookFavoriteChanges()
    }

    fun listenBookFavoriteChanges() {
        binding!!.favoriteBook.setOnClickListener { v: View? ->
            bookEntity!!.isFavorite = !bookEntity!!.isFavorite
            binding!!.favoriteBook.setImageDrawable(
                if (bookEntity!!.isFavorite) getDrawable(R.drawable.ic_heart_on) else getDrawable(R.drawable.ic_heart)
            )
            AppDatabase.getDatabase(applicationContext).bookDao.setBookAsFavoriteById(
                bookId,
                bookEntity!!.isFavorite
            )
        }
    }

    private fun setBookData(bookId: String) {
        bookEntity = AppDatabase.getDatabase(applicationContext).bookDao.getBookById(bookId)
        val bookEntity = bookEntity
        bookEntity ?: return
        formatAuthors(bookEntity.authors)
        binding!!.bookDescription.text = bookEntity.description
        binding!!.bookPageCount.text = getString(R.string.page_count_d, bookEntity.pageCount)
        binding!!.bookPublishedDate.text = bookEntity.publishedDate
        binding!!.bookTitle.text = bookEntity.title
        binding!!.bookRating.rating = bookEntity.averageRating.toFloat()
        binding!!.favoriteBook.setImageDrawable(
            if (bookEntity.isFavorite) getDrawable(R.drawable.ic_heart_on) else getDrawable(R.drawable.ic_heart)
        )
        Glide.with(this)
            .load(bookEntity.thumbnailUrl)
            .into(binding!!.bookImage)
    }

    private fun formatAuthors(authors: List<String>) {
        val authorsBuilder = StringBuilder()
        if (authors.isNotEmpty()) {
            for (author in authors) {
                if (authors.indexOf(author) == authors.size - 1) {
                    authorsBuilder.append(author)
                }
                authorsBuilder.append(String.format("%s, ", author))
            }
            binding!!.bookAuthors.text = authorsBuilder.toString()
        } else {
            binding!!.bookAuthors.setText(R.string.anonymous)
        }
    }
}